### ISSUE TITLE
Convert HEIC/HEIF images on upload from iOS

### ISSUE THREAD

**Issue by faxe78** [zulip/zulip-mobile#2178]
(Summary by @gnprice:)

Images created by recent versions of iOS are often in the HEIF format (extension `.heic`), which is a patent-restricted Apple format. When those images are uploaded, they tend to be unreadable elsewhere.

When uploading an image on iOS that's in HEIF format, we should use platform facilities to get or create a JPEG version and upload that.

(Original report follows.)

---

Hi,

I setup zulip today on a Ubuntu 16.04.4 LTS following the installation guide, so far everything looks promising. I found one issue so far or it looks like one at least.

Some pictures on my Iphone are JPG everything works, but some pictures are in HEIC format and here the latest IOS client only show a link like IMG_7113.HEIC no image, if i click the link a web view opens up loading but shows only a white page.

I see the request in the nginx logfile with a 200 status code.

xxxxxxx - - [30/Mar/2018:22:12:11 +0000] "GET /user_uploads/2/92/wPKjGi9juRThwf-r-0bW_HsO/IMG_7113.HEIC HTTP/1.1" 200 65247 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 11_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1"

I thought maybe for the web view it would help to add the mime types but that did not change anything.

/etc/nginx/zulip-include/uploads.types
    image/heic                            heic;
    image/heif                            heif;

Thx

Markus

**Comment by faxe78** [comment]
if i click on my desktop chrome on the image link the image is downloaded to my mac and preview can view this image.

I tried to upload the same picture in rocket chat and it uploads a jpg file for me it looks like i choose the same file. so it looks like the app can choose which format to upload or is it converted in the app? maybe that information helps.

**Comment by anirudhjain75** [comment]
@borisyankov I'd like to work on this

**Comment by anirudhjain75** [comment]
@faxe78 Could you send me a pic in heic or heir format ? Upload it to Drive and send me the link.

**Comment by faxe78** [comment]
looked and found an example image.

http://nokiatech.github.io/heif/content/images/autumn_1440x960.heic

there are more examples.

http://nokiatech.github.io/heif/examples.html

**Comment by anirudhjain75** [comment]
We could convert the image from heic/heif to commonly used format when uploading is done, This way is followed by other sharing tools like Telegram.

**Comment by gnprice** [comment]
Thanks @faxe78 for raising this!

> We could convert the image from heic/heif to commonly used format when uploading is done, This way is followed by other sharing tools like Telegram.

Yeah, we should upload a JPEG file. Unfortunately it looks like there isn't (and won't be) free software for handling HEIF images, because it's encumbered with a tangle of patents -- so we can't do the conversion on the server.

Based on some comments on https://github.com/ivpusic/react-native-image-crop-picker/issues/611 , it sounds like some tools may already provide both a JPEG and an HEIF file, and we just have to pick the right one.

Let's focus on just doing this on iOS -- there are likely to be platform-specific facilities for it, and it should only be a problem there anyway.

**Comment by gnprice** [comment]
@jainkuniya points out this project which aims to be a library we could use on the server: https://github.com/monostream/tifig

It's currently "in early alpha", so I wouldn't want to rely on it now. Given reports about the patent situation, I'm also skeptical that it can succeed. If at some point in the future it does and the patent situation seems to be OK, we might start using it on the server after all.

**Comment by zbenjamin** [comment]
https://github.com/react-community/react-native-image-picker/issues/872#issuecomment-404439605 (and https://github.com/ivpusic/react-native-image-crop-picker/issues/611) seem to imply that react-native-image-picker does provide a jpg. You just need to use `response.uri` instead of `response.fileName`. If I upload an HEIC file from the Zulip mobile app and then download it to my desktop, I see:

```
<12:46:52>windcharger ~/Downloads<0>$ file IMG_0722.HEIC 
IMG_0722.HEIC: JPEG image data, JFIF standard 1.01, aspect ratio, density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=2, orientation=upper-left], baseline, precision 8, 3024x4032, frames 3
```
so it does appear that the file is actually a JPEG.

It looks like `src/api/uploadFile.js` gets the extension/mimetype from `response.fileName` instead of `response.uri` so I think the fix for this is to just get the file name and extension/mimetype information from the URI, instead.

**Comment by jackrzhang** [comment]
I'm claiming this. (@zulipbot is disabled for this repository at the moment)

**Comment by jackrzhang** [comment]
A note on how to reproduce a bug for this:
1. On iOS, go to any message list.
2. Upload an image taken from within the Camera application - not from launching the camera from within Zulip. That image should be in the HEIC format.
3. A new message should be sent, and text for the image name with an extension for the HEIC format should appear. The photo itself isn't viewable within the message list.

**Comment by gnprice** [comment]
@faxe78 Thanks again for this detailed report! You, and others following this thread, may be glad to hear that version 17.1.98 went out to all users early this week, and has the fix.

**Comment by mohsinka** [comment]
@gnprice what did you do differently that made the HEIC image readable, did you fix it on the backend or compressed it on iOS?

**Comment by gnprice** [comment]
@mohsinka You can see the fix in the PR #2977 that closed the issue. The diagnosis was the one by @zbenjamin above: https://github.com/zulip/zulip-mobile/issues/2178#issuecomment-421467317 We were already uploading a JPEG file, and just needed to fix the filename to reflect that, because the Zulip server infers the intended format from the filename.
